sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ChannelLeaser.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import com.rabbitmq.client.Channel;
+import java.io.IOException;
+import java.util.UUID;
+
+/**
+ * A Channel is responsible for maintaining delivery state between client and server along a
+ * Connection. It's difficult to manage a Channel across different objects safely, so this 'loan
+ * pattern' is present to ensure safe, consistent access.
+ *
+ * <p>This interface uses the concept of a 'lessee', which is the identifier of a particular client,
+ * or set of clients, for a Channel. It is required that there be an "at-most-one" relationship
+ * between lessees and Channels; that is, any given id will have zero or one Channel associated with
+ * it.
+ */
+public interface ChannelLeaser {
+  /**
+   * A 'loan pattern' for utilizing a Channel such that the implementer handles exceptions and
+   * tracks the state of the Channel consistently, freeing the caller to exclusively worry about
+   * 'happy path' functionality.
+   */
+  @FunctionalInterface
+  interface UseChannelFunction<T> {
+    T apply(Channel channel) throws IOException;
```

Comment:
I think describing this as 'resource scoping' is a bit too limited as a description.

try-with-resources is also resource scoping but it's also inherently 'scoped' to the context of a function or call stack. This interface allows for resource 'borrowing' within the context of the lifetime of an application. 

I could implement this as `AutoCloseable` and allow for `Channel borrowChannel` function; but I don't think that would allow for consistent error handling and Channel/Connection state maintenance on error within the try-with-resources. 

The intent was to make it clear via the api that this is loaned and that error handling, Channel shutdown/recovery, resource pooling, would be handled by the system the Channel is being 'lent to' from. What I wanted to avoid was a situation where, say, the `CheckpointMark` and the `Reader` both internally store a reference to the same `Channel` and have to manage error handling and recovery together. That is, I did this as a means of data encapsulation of a 'precious resource'.

I am, however, feeling like this pattern is not well-liked by this community. I can see what alternatives might make sense here. Something like this maybe?

```java
interface ChannelResource extends AutoCloseable {
  Channel borrowChannel();

  void handleError(Exception e) throws Exception;
}

// using this in, say, Reader:
try (Channel channel = channelResource.borrowChannel()) {
  // do dangerous things with the Channel
} catch (Exception e) {
  channelResource.handleError(e);
}
// note: in the implicit `finally { channelHandler.close() }`
// the implementation should silently ignore state changes that may 
// have happened within `handleError` such as removing an unusable Channel 
// from the internal map
```